src/cargo/util/dependency_queue.rs

```diff
@@ -97,14 +101,53 @@ impl<K: Hash + Eq + Clone, V> DependencyQueue<K, V> {
         &mut slot.insert((my_dependencies, value)).1
     }
 
+    /// All nodes have been added, calculate some internal metadata and prepare
+    /// for `dequeue`.
+    pub fn queue_finished(&mut self) {
+        for key in self.dep_map.keys() {
+            depth(key, &self.dep_map, &mut self.depth, 0);
+        }
+
+        fn depth<K: Hash + Eq + Clone, V>(
+            key: &K,
+            map: &HashMap<K, (HashSet<K>, V)>,
+            results: &mut HashMap<K, usize>,
+            cur_depth: usize,
+        ) {
+            {
+                let prev_depth = results.entry(key.clone()).or_insert(0);
+                if *prev_depth > cur_depth {
+                    return
+                }
+                *prev_depth = cur_depth;
+            }
+
+            for dep in map[key].0.iter() {
+                depth(dep, map, results, cur_depth + 1);
+            }
+        }
+    }
```

Comment:
I believe this is O(N^2) worst case. Not sure if it matters, but I couldn't resist the temptation of implementing a dynamic programming algorithm :-)

```Rust
    pub fn queue_finished(&mut self) {
        for key in self.dep_map.keys() {
            depth(key, &self.reverse_dep_map, &mut self.depth);
        }

        fn depth<K: Hash + Eq + Clone>(
            key: &K,
            map: &HashMap<K, HashSet<K>>,
            results: &mut HashMap<K, usize>,
        ) -> usize {
            const IN_PROGRESS: usize = !0;

            if let Some(&depth) = results.get(key) {
                assert_ne!(depth, IN_PROGRESS, "cycle in DependencyQueue");
                return depth;
            }
            
            results.insert(key.clone(), IN_PROGRESS);

            let depth = map.get(&key).into_iter().flat_map(|it| it)
                .map(|dep| depth(dep, map, results))
                .max().unwrap_or(0) + 1;
            
            *results.get_mut(key).unwrap() = depth;
            
            depth
        }
    }
```

The idea is to use *reverse* map instead the usual one, calculate the depth as max of in-edges plus one and do a sanity check for absence of cycles just in case.